### ISSUE TITLE
Zookeeper operator fails to start pods due to headless service DNS issue

### ISSUE THREAD

**Issue by sphw** [loft-sh/loft#84]
I noticed an issue when trying to deploy pravega's zookeeper operator ( https://github.com/pravega/zookeeper-operator) in a virtual cluster. The operator installs fine, but when it tries to bring up the pods they fall into a crash loop with the following error: 

```
2021-02-02 06:57:17,626 [myid:1] - ERROR [QuorumPeer[myid=1](plain=0.0.0.0:2181)(secure=disabled):Leader@318] - Couldn't bind to zookeeper-0.zookeeper-headless.m10-zookeeper.svc.cluster.local:2888
java.net.SocketException: Unresolved address
	at java.base/java.net.ServerSocket.bind(Unknown Source)
	at java.base/java.net.ServerSocket.bind(Unknown Source)
	at org.apache.zookeeper.server.quorum.Leader.createServerSocket(Leader.java:315)
	at org.apache.zookeeper.server.quorum.Leader.lambda$new$0(Leader.java:294)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
	at java.base/java.util.concurrent.ConcurrentHashMap$KeySpliterator.forEachRemaining(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(Unknown Source)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline.forEach(Unknown Source)
	at org.apache.zookeeper.server.quorum.Leader.<init>(Leader.java:297)
	at org.apache.zookeeper.server.quorum.QuorumPeer.makeLeader(QuorumPeer.java:1260)
	at org.apache.zookeeper.server.quorum.QuorumPeer.run(QuorumPeer.java:1467)
```

I did some digging, and it seems like the root cause is a problem with DNS for headless services in a virtual cluster. The DNS entry for `zookeeper-0.zookeeper-headless.m10-zookeeper.svc.cluster.local` is not being created. I suspect this is because pods are renamed when they are synced into the host-cluster, and the host cluster control plane is responsible for service creation.

Any help with this issue would be appreciated.

**Comment by FabianKramm** [comment]
@sphw thanks for reporting this issue! We are aware of this issue and it will be fixed in the next version. Some background information on why this happens:

When you create a statefulset in kubernetes, the created pods of the statefulset will have the `spec.subdomain` option set to the name of the headless service. This subdomain will basically create an entry of the form `POD_NAME.SUBDOMAIN.NAMESPACE.svc.cluster.local` in the `/etc/hosts` file of a container. If you run `hostname -f` it will yield the above domain. This is so far normal kubernetes behaviour. 

In the virtual cluster however, we cannot forward the subdomain option to the physical cluster, because that would lead to a wrong `/etc/hosts` entry since kubernetes would write the actual physical namespace in there instead of the virtual cluster namespace. Thats why we currently just empty that option and pass it to the physical cluster. This is usually not a problem, except for applications that specifically need that fully qualified hostname inside the container and either want to bind to it or want to find it out via `hostname -f`. In this case such problems as you experience occur.

In the next version we will override the `/etc/hosts` file for each container in a pod that specifies the `spec.subdomain` option automatically with the correct entries in an init container, which should fix that issue.

**Comment by FabianKramm** [comment]
@sphw this issue should be fixed with loft version v1.7.0! Make sure you create a new virtual cluster for this to take effect.
